File: daemon/src/services/githubAuth.ts
/**
 * GitHub credential resolution (D2/D3 in
 * `.vibekit/reports/2026-08-16-pr-detection-broken-root-cause-and-fix.md`).
 *
 * Deliberately does NOT require the `gh` binary — `gh` is not provisioned by
 * `dev.Dockerfile` (or any environment this daemon runs in), so it can only
 * ever be an opportunistic last resort, never a hard dependency. Reads gh's
 * own *data file* (`~/.config/gh/hosts.yml`) instead, which holds a
 * per-login plaintext `oauth_token` on every host inspected so far (a
 * keyring-backed `hosts.yml` with no plaintext token is a known, unhandled
 * gap — see the research report's "Not checked" section).
 *
 * Credential chain, per account login (D3):
 *   1. `GH_TOKEN_<LOGIN>` env var (login upper-cased, `-`/`.` -> `_`)
 *   2. `GITHUB_TOKEN` / `GH_TOKEN` env var (generic — single-account fallback)
 *   3. `~/.config/gh/hosts.yml` -> `users.<login>.oauth_token`
 *   4. `gh auth token --user <login>`, only if `gh` happens to be on PATH
 *
 * `listAccounts()` never throws; it resolves to `[]` when nothing is found.
 */
import { readFile } from "node:fs/promises";
import { homedir } from "node:os";
import { join } from "node:path";
import { execFile as execFileCb } from "node:child_process";
import { promisify } from "node:util";

const execFile = promisify(execFileCb);

export interface GithubAccount {
  login: string;
  token: string;
}

function hostsYmlPath(): string {
  return join(process.env.GH_CONFIG_DIR || join(homedir(), ".config", "gh"), "hosts.yml");
}

/**
 * Minimal, purpose-built parser for gh's `hosts.yml` shape — not a general
 * YAML parser (no library dependency for one file format). Handles:
 *
 *   github.com:
 *       users:
 *           some-login:
 *               oauth_token: gho_xxxx
 *           other-login: {}
 *       oauth_token: gho_xxxx   # top-level = currently-active user's token
 *       user: some-login
 *
 * Returns every login found under the `github.com:` host's `users:` block
 * only — logins nested under a GitHub Enterprise Server host block (a
 * top-level key other than `github.com`) are deliberately excluded, since
 * their tokens are only valid against that GHES instance's API, not
 * `api.github.com`, and this daemon only ever talks to the latter.
 *
 * `oauth_token` when present (absent for keyring-backed logins).
 */
export function parseHostsYml(text: string): Array<{ login: string; token: string | null }> {
  const lines = text.split("\n");
  const found: Array<{ login: string; token: string | null }> = [];

  let currentHost: string | null = null;
  let inUsers = false;
  let usersIndent = -1;
  let currentLogin: string | null = null;
  let currentLoginIndent = -1;
  let currentToken: string | null = null;

  const flushCurrent = () => {
    if (currentLogin !== null && currentHost === "github.com") {
      found.push({ login: currentLogin, token: currentToken });
    }
    currentLogin = null;
    currentToken = null;
  };

  for (const raw of lines) {
    if (!raw.trim() || raw.trim().startsWith("#")) continue;
    const indent = raw.length - raw.trimStart().length;
    const trimmed = raw.trim();

    // Top-level host block header, e.g. `github.com:` or a GHES hostname.
    if (indent === 0) {
      const hostMatch = /^([^:]+):\s*$/.exec(trimmed);
      if (hostMatch) {
        flushCurrent();
        inUsers = false;
        currentHost = hostMatch[1]!.trim();
        continue;
      }
    }

    if (trimmed === "users:") {
      flushCurrent();
      inUsers = true;
      usersIndent = indent;
      continue;
    }
    if (!inUsers) continue;

    if (indent <= usersIndent) {
      flushCurrent();
      inUsers = false;
      continue;
    }

    if (currentLogin === null || indent <= currentLoginIndent) {
      // Strip a trailing ` # comment` before matching — `login: # nickname`
      // is valid hosts.yml and previously failed to match at all, silently
      // dropping that login.
      const withoutComment = trimmed.replace(/\s+#.*$/, "");
      const m = /^([^:]+):\s*(\{\s*\})?\s*$/.exec(withoutComment);
      if (m) {
        flushCurrent();
        currentLogin = m[1]!.trim();
        currentLoginIndent = indent;
      }
      continue;
    }

    const tokenMatch = /^oauth_token:\s*(.+)$/.exec(trimmed);
    if (tokenMatch) {
      currentToken = tokenMatch[1]!.trim().replace(/^["']|["']$/g, "");
    }
  }
  flushCurrent();
  return found;
}

async function readHostsYmlAccounts(): Promise<Array<{ login: string; token: string | null }>> {
  try {
    const text = await readFile(hostsYmlPath(), "utf8");
    return parseHostsYml(text);
  } catch {
    return [];
  }
}

async function isGhOnPath(): Promise<boolean> {
  try {
    await execFile("gh", ["--version"]);
    return true;
  } catch {
    return false;
  }
}

async function ghAuthTokenForUser(login: string): Promise<string | null> {
  try {
    const { stdout } = await execFile("gh", ["auth", "token", "--user", login]);
    const token = stdout.trim();
    return token || null;
  } catch {
    return null;
  }
}

function envTokenForLogin(login: string): string | undefined {
  const key = `GH_TOKEN_${login.toUpperCase().replace(/[-.]/g, "_")}`;
  return process.env[key];
}

/**
 * Resolves every GitHub account this daemon can act as, one token per login,
 * strongest source wins per the chain above. Cheap enough to call per poll
 * tick (one file read, no network) that no in-memory cache is needed here —
 * callers that want a stable view across a tick should call it once and
 * reuse the result.
 */
export async function listAccounts(): Promise<GithubAccount[]> {
  const tokens = new Map<string, string>();
  const knownLogins = new Set<string>();

  // 3. hosts.yml — establishes known logins even when a login has no
  //    plaintext token (keyring-backed), so step 4 still knows to try it.
  const hostsAccounts = await readHostsYmlAccounts();
  for (const { login, token } of hostsAccounts) {
    knownLogins.add(login);
    if (token) tokens.set(login, token);
  }

  // 2a. Generic GITHUB_TOKEN/GH_TOKEN — env wins over hosts.yml, but ONLY when
  //    there's a single account in play (covers the common solo-account case
  //    with no hosts.yml at all — synthetic "default" login). With >=2 known
  //    logins a single generic token cannot possibly be valid for all of
  //    them, so filling it in here is deferred to 2b below, AFTER step 4 has
  //    had a chance to resolve each login's own real token — see 2b for why.
  const generic = process.env.GITHUB_TOKEN || process.env.GH_TOKEN;
  if (generic) {
    if (knownLogins.size === 0) {
      knownLogins.add("default");
      tokens.set("default", generic);
    } else if (knownLogins.size === 1) {
      for (const login of knownLogins) tokens.set(login, generic);
    }
  }

  // 1. GH_TOKEN_<LOGIN> — highest priority, per-account override.
  for (const login of knownLogins) {
    const perAccount = envTokenForLogin(login);
    if (perAccount) tokens.set(login, perAccount);
  }

  // 4. `gh auth token --user X` — only for logins still missing a token, and
  //    only if `gh` happens to be on PATH. Never required. Must run BEFORE
  //    2b below: a keyring-backed second login has no plaintext token in
  //    hosts.yml, so if 2b's generic-token fill ran first it would stamp the
  //    (wrong-for-this-login) generic token onto it and this step would then
  //    see it as already resolved and skip it — the exact wrong-token->404
  //    failure this fallback chain exists to avoid.
  const stillMissing = [...knownLogins].filter((login) => !tokens.has(login));
  if (stillMissing.length > 0 && (await isGhOnPath())) {
    for (const login of stillMissing) {
      const token = await ghAuthTokenForUser(login);
      if (token) tokens.set(login, token);
    }
  }

  // 2b. Generic GITHUB_TOKEN/GH_TOKEN, >=2-account fallback — only NOW, for
  //    logins that even `gh auth token --user` couldn't resolve a real token
  //    for. Still a guess (one token for every remaining login), but it's a
  //    last resort after every login-specific source has been tried, not a
  //    pre-emptive stamp that blocks a better source from running.
  if (generic && knownLogins.size >= 2) {
    for (const login of knownLogins) {
      if (!tokens.has(login)) tokens.set(login, generic);
    }
  }

  return [...tokens.entries()].map(([login, token]) => ({ login, token }));
}

File: rust/vst-lifecycle/src/github_auth.rs
//! GitHub credential resolution — ports `services/githubAuth.ts`.
//!
//! Behavior contract (D2/D3 — see TS source header):
//! - Never requires the `gh` binary (not provisioned in CI).
//! - Credential chain (D3):
//!   1. `GH_TOKEN_<LOGIN>` env var per login (LOGIN uppercased)
//!   2. `GITHUB_TOKEN` / `GH_TOKEN` generic env vars — fills logins with no
//!      token only when there is exactly one known login (step 2b: a generic
//!      token only fills logins with no token AFTER step 4, but only if there
//!      is no ambiguity — applies to the single-account case)
//!   3. `~/.config/gh/hosts.yml` (or `$GH_CONFIG_DIR/hosts.yml`)
//!   4. `gh auth token --user X` only if `gh` happens to be on PATH
//! - Never panics; returns `[]` on any failure.

use std::process::Command;

/// A GitHub account credential resolved from the chain.
#[derive(Clone, Debug, PartialEq)]
pub struct GithubAccount {
    pub login: String,
    pub token: Option<String>,
}

/// Parse a `~/.config/gh/hosts.yml`-shaped file for `github.com` credentials.
///
/// Only the `github.com:` block is processed; any GHES blocks are excluded.
/// A login with no plaintext `oauth_token` (e.g. keyring-backed) returns
/// `token: None`.
pub fn parse_hosts_yml(text: &str) -> Vec<GithubAccount> {
    let mut result = Vec::new();
    let mut in_github_com = false;
    let mut in_users = false;
    let mut current_login: Option<String> = None;

    for line in text.lines() {
        let trimmed = line.trim_end();

        // Top-level host key (zero-indent, ends with ':')
        if !trimmed.starts_with(' ') && !trimmed.starts_with('\t') {
            let key = trimmed.trim_end_matches(':');
            in_github_com = key == "github.com";
            in_users = false;
            current_login = None;
            continue;
        }

        if !in_github_com {
            continue;
        }

        let stripped = trimmed.trim_start();
        let indent = trimmed.len() - stripped.len();

        // `    users:` — 4-space indent
        if stripped == "users:" {
            in_users = true;
            current_login = None;
            continue;
        }

        if !in_users {
            continue;
        }

        // login lines are at 8-space indent.
        // Formats: `alice:`, `keyring-user: {}`, `commented-login: # comment`
        if indent == 8 && stripped.contains(':') {
            let colon_pos = stripped.find(':').unwrap();
            let login_candidate = stripped[..colon_pos].trim();
            if !login_candidate.is_empty() && !login_candidate.contains(' ') {
                current_login = Some(login_candidate.to_string());
                result.push(GithubAccount {
                    login: login_candidate.to_string(),
                    token: None,
                });
            }
            continue;
        }

        // `oauth_token: <value>` lines are at 12-space indent
        if indent == 12 && stripped.starts_with("oauth_token:") {
            if let Some(ref login) = current_login {
                let token_val = stripped["oauth_token:".len()..].trim();
                if !token_val.is_empty() {
                    if let Some(entry) = result.iter_mut().find(|a| &a.login == login) {
                        entry.token = Some(token_val.to_string());
                    }
                }
            }
        }
    }

    result
}

fn gh_config_dir() -> Option<std::path::PathBuf> {
    if let Some(d) = std::env::var_os("GH_CONFIG_DIR") {
        return Some(std::path::PathBuf::from(d));
    }
    // Default: ~/.config/gh — use HOME env var (POSIX standard, always set on Linux).
    std::env::var_os("HOME").map(|h| std::path::PathBuf::from(h).join(".config").join("gh"))
}

/// Resolve GitHub accounts using the full credential chain.
/// Returns an empty vec on any failure; never panics.
pub async fn list_accounts() -> Vec<GithubAccount> {
    // Step 3: parse hosts.yml to get the known logins.
    let mut accounts: Vec<GithubAccount> = {
        let config_dir = gh_config_dir();
        let hosts_path = config_dir.map(|d| d.join("hosts.yml"));
        if let Some(path) = hosts_path {
            if let Ok(text) = tokio::fs::read_to_string(&path).await {
                parse_hosts_yml(&text)
            } else {
                vec![]
            }
        } else {
            vec![]
        }
    };

    // Step 4: try `gh auth token --user <login>` for logins still missing tokens.
    for acc in accounts.iter_mut() {
        if acc.token.is_some() {
            continue;
        }
        if let Ok(output) = Command::new("gh")
            .args(["auth", "token", "--user", &acc.login])
            .output()
        {
            if output.status.success() {
                let token = String::from_utf8_lossy(&output.stdout).trim().to_string();
                if !token.is_empty() {
                    acc.token = Some(token);
                }
            }
        }
    }

    // Step 1: per-login env var overrides (GH_TOKEN_<LOGIN>, uppercased).
    for acc in accounts.iter_mut() {
        let env_key = format!("GH_TOKEN_{}", acc.login.to_ascii_uppercase());
        if let Ok(token) = std::env::var(&env_key) {
            if !token.is_empty() {
                acc.token = Some(token);
            }
        }
    }

    // Step 2b: generic GITHUB_TOKEN / GH_TOKEN fills only logins that still
    // have no token, but only when there's exactly ONE known login (no
    // ambiguity across accounts — a generic token for the wrong account would
    // silently 404 on that account's private repos).
    let generic = std::env::var("GITHUB_TOKEN")
        .ok()
        .filter(|t| !t.is_empty())
        .or_else(|| std::env::var("GH_TOKEN").ok().filter(|t| !t.is_empty()));

    if let Some(generic_token) = generic {
        if accounts.len() == 1 {
            if let Some(acc) = accounts.iter_mut().find(|a| a.token.is_none()) {
                acc.token = Some(generic_token);
            }
        }
    }

    accounts
}
